File: Leetcode_Python/copy_list_with_random_pointer.py
# Definition for singly-linked list with a random pointer.
# class RandomListNode:
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None

class Solution:
    # @param head, a RandomListNode
    # @return a RandomListNode
    def copyRandomList(self, head):
        if not head:
            return None
        chead = RandomListNode(head.label)
        d = {head: chead}
        p = chead
        node = head.next
        while node:
            tem = RandomListNode(node.label)
            d[node] = tem
            p.next = tem
            p = tem
            node = node.next
        
        d[None] = None
        node = head
        q = chead
        while node:
            q.random = d[node.random]
            node = node.next
            q = q.next
       
        return chead

File: cleanCode50/CopyListWithRandomPointer.java
class RandomListNode {
    int label;
    RandomListNode next;
    RandomListNode random;
    RandomListNode(int x) { 
        this.label = x; 
    }
}

public class CopyListWithRandomPointer {
    public RandomListNode copyRandomList(RandomListNode head) {
        if (head == null) {
            return head;
        }
        RandomListNode headCopy = new RandomListNode(head.label);
        Map<RandomListNode, RandomListNode> map = new HashMap<>();
        map.put(head, headCopy);
        RandomListNode p = head;
        while (p != null && (p.next != null || p.random != null)) {
            if (p.next != null) {
                if (!map.containsKey(p.next)) {
                    RandomListNode nextNode = new RandomListNode(p.next.label);
                    map.put(p.next, nextNode);
                }
                map.get(p).next = map.get(p.next);
            }
            if (p.random != null) {
                if (!map.containsKey(p.random)) {
                    RandomListNode randomNode = new RandomListNode(p.random.label);
                    map.put(p.random, randomNode);
                }
                map.get(p).random = map.get(p.random);
            }
            p = p.next;
        }
        return headCopy;
    }
}